lib/stack_master/aws_driver/s3.rb

```diff
@@ -0,0 +1,66 @@
+require 'digest/md5'
+
+module StackMaster
+  module AwsDriver
+    class S3ConfigurationError < StandardError; end
+
+    class S3
+      def set_region(region)
+        @region = region
+        @s3 = nil
+      end
+
+      def upload_files(bucket: nil, prefix: nil, region: nil, files: {})
+        raise StackMaster::AwsDriver::S3ConfigurationError, 'A bucket must be specified in order to use S3' unless bucket
+
+        s3 = new_s3_client(region: region)
+
+        return if files.empty?
+
+        current_objects = s3.list_objects(
+          prefix: prefix,
+          bucket: bucket
+        ).map(&:contents).flatten.inject({}){|h,obj|
+          h.merge(obj.key => obj)
+        }
+
+        StackMaster.stdout.puts "Uploading files to S3:"
+
+        files.each do |template, file|
+          body = file.fetch(:body)
+          path = file.fetch(:path)
+          object_key = template.dup
+          object_key.prepend("#{prefix}/") if prefix
+          compiled_template_md5 = Digest::MD5.hexdigest(body).to_s
+          s3_md5 = current_objects[object_key] ? current_objects[object_key].etag.gsub("\"", '') : nil
+
+          next if compiled_template_md5 == s3_md5
+          s3_uri = "s3://#{bucket}/#{object_key}"
+          StackMaster.stdout.print "- #{File.basename(path)} => #{s3_uri} "
+
+          s3.put_object(
```

Comment:
This won't work for large files. Should we raise an exception early? Should we document this limitation?